### DISCUSSION TITLE
How to use `signAuthMessage` for eip3074 without having to pass private key of the authority?

### DISCUSSION THREAD

**Discussion by swetshaw** [wevm/viem#2199] [Question]
I want the authority to come to my platform (website), authorise my invoker contract (sign the digest using metamask) to invoke the txs on their behalf. The invoker contract at any later time could execute the tx. 

Major issue that I am facing here is - I am unable to find a way for the authority to sign the commit without having to pass the private key. Do you know if there is a way to do it?

**Comment by jxom** [comment]
There isn't a first-class abstraction to do this yet, as users shouldn't sign raw digests via their wallets (very dangerous) – unless a wallet were to eventually expose a `wallet_signAuthMessage` method (akin to `eth_signTransaction`). 

However, if you want to turn on `eth_sign` in MetaMask, I can export the [`toAuthMessage` utility](https://github.com/wevm/viem/blob/main/src/experimental/eip3074/utils/toAuthMessage.ts) so you can perform a raw request to sign the auth message:

```ts
import { toAuthMessage } from 'viem/experimental'
import { client } from './config'

const signature = await client.request({ 
  method: 'eth_sign',
  parameters: [keccak256(toAuthMessage({ chainId, commit, nonce, invokerAddress }))]
})
```

Would that work?

If you already have the digest, you can also do:

```ts
import { toAuthMessage } from 'viem/experimental'
import { client } from './config'

const digest = '0x...'

const signature = await client.request({ 
  method: 'eth_sign',
  parameters: [digest]
})
```

**Reply by swetshaw** [reply]
I see and that absolutely makes sense. This makes me wonder how is it supposed to behave in an ideal world, assuming authority shouldn't provide their private key for the auth. Do you know what's the ideal flow for this?

**Reply by swetshaw** [reply]
Exporting `toAuthMessage` utility will be helpful, I can utilise it for a quick demo that I am working on
